File: January/January 17/cow_checkups.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

int main() {
    int number_of_cows;
    std::cin >> number_of_cows;

    std::vector<int> cow_line(number_of_cows);
    std::vector<int> wanted_cow_line(number_of_cows);

    for (int i = 0; i < number_of_cows; ++i) {
        std::cin >> cow_line[i];
    }
    for (int i = 0; i < number_of_cows; ++i) {
        std::cin >> wanted_cow_line[i];
    }

    std::unordered_map<int, int> answers;

    int start = 0;
    for (int cow_index = 0; cow_index < number_of_cows; ++cow_index) {
        if (wanted_cow_line[cow_index] == cow_line[cow_index]) {
            ++start;
        }
    }

    for (int l = 0; l < number_of_cows; ++l) {
        for (int r = l; r < number_of_cows; ++r) {
            int answer = start;
            std::vector<int> cow_line_copy = cow_line;

            for (int cow_index = l; cow_index <= r; ++cow_index) {
                if (wanted_cow_line[cow_index] == cow_line_copy[cow_index]) {
                    --answer;
                }
            }

            std::reverse(cow_line_copy.begin() + l, cow_line_copy.begin() + r + 1);

            for (int cow_index = l; cow_index <= r; ++cow_index) {
                if (wanted_cow_line[cow_index] == cow_line_copy[cow_index]) {
                    ++answer;
                }
            }

            ++answers[answer];
        }
    }

    for (int number = 0; number <= number_of_cows; ++number) {
        std::cout << answers[number] << "\n";
    }

    return 0;
}

File: January/January 17/cow_checkups.py
# https://usaco.org/index.php?page=viewproblem2&cpid=1469

from collections import defaultdict

number_of_cows = int(input())
cow_line = list(map(int,input().split(' ')))
wanted_cow_line = list(map(int,input().split(' ')))

answers = defaultdict(int)

start = 0

for cow_index in range(number_of_cows):
    if wanted_cow_line[cow_index] == cow_line[cow_index]:
        start += 1

answers[start] = number_of_cows

for l in range(0,number_of_cows):
    for r in range(l+1,number_of_cows):

        answer = start

        cow_line_copy = cow_line.copy()

        for cow_index in range(l,r+1):
            if wanted_cow_line[cow_index] == cow_line_copy[cow_index]:
                answer -= 1

        cow_line_copy[l:r+1] = cow_line_copy[l:r+1][::-1]

        for cow_index in range(l,r+1):
            if wanted_cow_line[cow_index] == cow_line_copy[cow_index]:
                answer += 1

        answers[answer] += 1

for number in range(number_of_cows+1):
    print(answers[number])